pkg/admission/server.go

```diff
@@ -0,0 +1,180 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	admission "k8s.io/api/admission/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/klog/v2"
+
+	v1alpha1 "sigs.k8s.io/service-apis/apis/v1alpha1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func log500(w http.ResponseWriter, err error) {
+	klog.Errorf("failed to read request from client: %v\n", err)
+	http.Error(w, err.Error(), http.StatusInternalServerError)
+	return
+}
+
+// ensureKindAdmissionReview check that our admission server is only getting requests
+// for kind AdmissionReview and reject all others
+func ensureKindAdmissionReview(req []byte) bool {
+	type reqBody struct {
+		Kind  string                 `json:"kind"`
+		Extra map[string]interface{} `json:"-"`
+	}
+	var msg reqBody
+	err := json.Unmarshal(req, &msg)
+	if err != nil {
+		return false
+	}
+	if msg.Kind != "AdmissionReview" {
+		return false
+	}
+	return true
+}
+
+// ServeHTTP parses AdmissionReview requests and responds back
+// with the validation result of the entity.
+func ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		klog.Errorf("webhook cannot handle %s method", r.Method)
+		http.Error(w, fmt.Sprintf("invalid method %s, only POST requests are allowed", r.Method), http.StatusMethodNotAllowed)
+	}
+	if r.Body == nil {
+		klog.Errorf("received request with empty body")
+		http.Error(w, "admission review object is missing",
+			http.StatusBadRequest)
+		return
+	}
+	data, readErr := ioutil.ReadAll(r.Body)
+	if readErr != nil {
+		log500(w, readErr)
+	}
+	review := admission.AdmissionReview{}
+	if err := json.Unmarshal(data, &review); err != nil {
+		klog.Errorf("failed to parse AdmissionReview object: %v\n", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	if !ensureKindAdmissionReview(data) {
+		invalidKind := "submitted object is not of kind AdmissionReview"
+		klog.Errorf(invalidKind)
+		http.Error(w, invalidKind, http.StatusBadRequest)
+		return
+	}
+	response, err := handleValidation(*review.Request)
+	if err != nil {
+		log500(w, err)
+	}
+	review.Response = response
+	data, err = json.Marshal(review)
+	if err != nil {
+		log500(w, err)
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		log500(w, err)
+	}
+	return
+}
+
+var (
+	httpRoute = meta.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  v1alpha1.SchemeGroupVersion.Version,
+		Resource: "httproutes",
```

Comment:
Ambivalent about this. We should consider defining this as a constant in the `api` package instead as this will be needed in other places.

We could also consider defining GVR in the api package itself.

Just a thought, doesn't need to happen in this PR.